File: 3163_string_compression_iii.cpp
class Solution { 
public:
    static string compressedString(string& s) {
        int len=s.size(), i=0;
        string out;
        out.reserve(len);
        for(int j=0; j<len; ) {
            while(j<len && s[j]==s[i]) j++;
            auto [parts, left]=div(j-i, 9);
            for(int k=0; k<parts; k++) {
                out+='9';
                out+=s[i];
            }
            if(left>0) {
                out+=left+'0';
                out+=s[i];
            }
            i=j;
        }
        return out;
    }
};

File: 3163_string_compression_iii.java
class Solution {
    public String compressedString(String str) {
        int size = str.length(), start = 0;
        StringBuilder result = new StringBuilder(size);
        for (int end = 0; end < size; ) {
            while (end < size && str.charAt(end) == str.charAt(start)) end++;
            int q = (end - start) / 9, r = (end - start) % 9;
            for (int c = 0; c < q; c++) result.append("9").append(str.charAt(start));
            if (r > 0) result.append(r).append(str.charAt(start));
            start = end;
        }
        return result.toString();
    }
}
